File: packages/rules-store/src/store.ts
/**
 * Versioned legal rules/config store.
 *
 * Never hardcode a legal rule (CLAUDE.md): AfA rates, HKVO ratios, the CO₂ 10-step
 * table, Anlage-V line numbers etc. are DATA with an "as-of law date". Engines ask
 * for a rule as of a date and stamp `Rechtsstand MM/JJJJ` into every legal output.
 */

export interface RuleVersion<T> {
  /** The date this version of the rule became law ("Rechtsstand"). ISO date. */
  readonly validFrom: string;
  /** Optional end of validity (exclusive). Open-ended if absent. */
  readonly validTo?: string;
  /** Citation, e.g. "§ 7 Abs. 1 HeizkostenV". */
  readonly source: string;
  readonly value: T;
}

export interface RuleSet<T> {
  readonly key: string;
  readonly versions: readonly RuleVersion<T>[];
}

export interface ResolvedRule<T> {
  readonly key: string;
  readonly value: T;
  readonly source: string;
  /** e.g. "Rechtsstand 12/2023" — must be shown in every legal output. */
  readonly rechtsstand: string;
}

export class RuleNotFoundError extends Error {
  constructor(key: string, asOf: string) {
    super(`No version of rule "${key}" is valid as of ${asOf}`);
    this.name = 'RuleNotFoundError';
  }
}

export function formatRechtsstand(isoDate: string): string {
  const [year, month] = isoDate.split('-');
  if (!year || !month) {
    throw new Error(`Invalid ISO date for Rechtsstand: ${isoDate}`);
  }
  return `Rechtsstand ${month}/${year}`;
}

/**
 * Resolves the version of a rule valid as of `asOf` (ISO date): the latest
 * `validFrom <= asOf` whose `validTo` (if any) is still open at `asOf`.
 */
export function getRule<T>(ruleSet: RuleSet<T>, asOf: string): ResolvedRule<T> {
  const applicable = [...ruleSet.versions]
    .filter((v) => v.validFrom <= asOf && (!v.validTo || asOf < v.validTo))
    .sort((a, b) => (a.validFrom < b.validFrom ? 1 : -1));

  const match = applicable[0];
  if (!match) {
    throw new RuleNotFoundError(ruleSet.key, asOf);
  }
  return {
    key: ruleSet.key,
    value: match.value,
    source: match.source,
    rechtsstand: formatRechtsstand(match.validFrom),
  };
}

File: packages/rules-store/src/lokara_rules_store/store.py
"""Versioned legal rules/config store.

Never hardcode a legal rule (CLAUDE.md): AfA rates, HKVO ratios, the CO₂
10-step table, Anlage-V line numbers etc. are DATA with an "as-of law date".
Engines receive a resolved rule as a parameter and stamp `Rechtsstand MM/JJJJ`
into every legal output.
"""

from dataclasses import dataclass
from datetime import date


@dataclass(frozen=True)
class RuleVersion[T]:
    # The date this version of the rule became law ("Rechtsstand").
    valid_from: date
    # Citation, e.g. "§ 7 Abs. 1 HeizkostenV".
    source: str
    value: T
    # Optional end of validity (exclusive). Open-ended if None.
    valid_to: date | None = None


@dataclass(frozen=True)
class RuleSet[T]:
    key: str
    versions: tuple[RuleVersion[T], ...]


@dataclass(frozen=True)
class ResolvedRule[T]:
    key: str
    value: T
    source: str
    # e.g. "Rechtsstand 12/2023" — must be shown in every legal output.
    rechtsstand: str


class RuleNotFoundError(LookupError):
    def __init__(self, key: str, as_of: date) -> None:
        super().__init__(f'No version of rule "{key}" is valid as of {as_of.isoformat()}')


def format_rechtsstand(as_of: date) -> str:
    return f"Rechtsstand {as_of.month:02d}/{as_of.year}"


def get_rule[T](rule_set: RuleSet[T], as_of: date) -> ResolvedRule[T]:
    """Resolves the version of a rule valid as of ``as_of``: the latest
    ``valid_from <= as_of`` whose ``valid_to`` (if any) is still open at ``as_of``.
    """
    applicable = [
        v
        for v in rule_set.versions
        if v.valid_from <= as_of and (v.valid_to is None or as_of < v.valid_to)
    ]
    if not applicable:
        raise RuleNotFoundError(rule_set.key, as_of)
    match = max(applicable, key=lambda v: v.valid_from)
    return ResolvedRule(
        key=rule_set.key,
        value=match.value,
        source=match.source,
        rechtsstand=format_rechtsstand(match.valid_from),
    )
